src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -6744,6 +6745,208 @@ def merge(
+            If False, don’t match the same ‘on’ value (i.e., strictly less-than / strictly greater-than).
+        direction : ‘backward’ (default), ‘forward’, or ‘nearest’
+            Whether to search for prior, subsequent, or closest matches.
+
+        Returns
+        -------
+        SnowflakeQueryCompiler
+        """
+        if (
+            by
+            or left_by
+            or right_by
+            or left_index
+            or right_index
+            or tolerance
+            or suffixes != ("_x", "_y")
+        ):
+            ErrorMessage.not_implemented(
+                "Snowpark pandas merge_asof method does not currently support parameters "
+                + "'by', 'left_by', 'right_by', 'left_index', 'right_index', "
+                + "'suffixes', or 'tolerance'"
+            )
+        if direction not in ("backward", "forward"):
+            ErrorMessage.not_implemented(
+                "Snowpark pandas merge_asof method only supports directions 'forward' and 'backward'"
+            )
+
+        left_frame = self._modin_frame
+        right_frame = right._modin_frame
+        # Project to Snowpark DataFrame in order to perform ASOF Join
+        left_snowpark_df = (
+            left_frame.ordered_dataframe.to_projected_snowpark_dataframe()
+        )
+        right_snowpark_df = (
+            right_frame.ordered_dataframe.to_projected_snowpark_dataframe()
+        )
+        if on:
+            left_on_quoted_identifier = (
+                left_frame.get_snowflake_quoted_identifiers_group_by_pandas_labels(
+                    [on]
+                )[0][0]
+            )
+            right_on_quoted_identifier = (
+                right_frame.get_snowflake_quoted_identifiers_group_by_pandas_labels(
+                    [on]
+                )[0][0]
+            )
+        else:
+            assert left_on and right_on
+            left_on_quoted_identifier = (
+                left_frame.get_snowflake_quoted_identifiers_group_by_pandas_labels(
+                    [left_on]
+                )[0][0]
+            )
+            right_on_quoted_identifier = (
+                right_frame.get_snowflake_quoted_identifiers_group_by_pandas_labels(
+                    [right_on]
+                )[0][0]
+            )
+
+        lhs_expr = left_snowpark_df[left_on_quoted_identifier]
+        rhs_expr = right_snowpark_df[right_on_quoted_identifier]
+        order_by_expr = [
+            left_snowpark_df[left_on_quoted_identifier],
+            right_snowpark_df[right_on_quoted_identifier],
+        ]
+        if direction == "backward":
+            match_condition = (
+                (lhs_expr >= rhs_expr) if allow_exact_matches else (lhs_expr > rhs_expr)
+            )
+        else:
+            match_condition = (
+                (lhs_expr <= rhs_expr) if allow_exact_matches else (lhs_expr < rhs_expr)
+            )
+
+        joined_snowpark_df = left_snowpark_df.join(
```

Comment:
snowpark pandas is not suppose to be directly operate on snowpark dataframe, all operation has to go through ordered dataframe, please make sure we extend our join_utils, and join operator at ordered dataframe properly
cc @sfc-gh-nkumar please also help review this pr to help make sure we extend the utility properly